Spec: customizing clients

chrisradek opened the pull request:
Using this PR to help collect scenarios that drive `client.tsp` customizations.  Related to #12183

Current status:
Lacking enough scenarios that are triggered by some automatable process for driving client.tsp updates. As such, the value proposition for creating a language-agnostic workflow is relatively low given we already have tooling be written for .NET, and a prompt for Go's scenario. Need input from other language teams to tip the scales!

Commit:
draft of customizing-client-tsp spec

maririos commented on tools/azsdk-cli/docs/specs/3-customizing-client-tsp.spec.md:
```diff
@@ -0,0 +1,93 @@
+
+Currently, feedback for adding client customizations to `client.tsp` can come in 2 ways:
+
+1. From manual reviews, e.g. API View feedback
```
Consider modifying it to `From reviews` because they can come from APIView Copilot (not manual :) )

maririos reviewed (commented):
Thanks for putting this together. General feedback:
-  I see the document doesn't talk about customizations to the SDK per se. Could you please update it and make sure it covers that case too?
- From my understanding, customization happens when there is a trigger saying a change is needed. The trigger then follows a process. This document should identify the triggers that we know of: build failure (caused by analyzers or breaking changes), feedback received by either people or tooling, and for user experience (breaking changes). With each trigger we can look: how to identify a change needs to happen? do we need to classify it to see how to address the customization? what is the success criteria?
- Once the triggers are identified, then what is the process to address them. According to the trigger the process can be different(?), but hopefully they all use the same steps/base. This is the part that is more confusing and that needs more time. Building the bones for that will be key in this doc

lmazuel commented on tools/azsdk-cli/docs/specs/3-customizing-client-tsp.spec.md:
```diff
@@ -0,0 +1,93 @@
+
+- **<a id="typespec"></a>TypeSpec**: The specification language used to define Azure service APIs. SDK code is generated from TypeSpec specifications located in the azure-rest-api-specs repository.
+
+- **<a id="typespec-customizations"></a>TypeSpec Customizations**: SDK-specific customizations made in the `client.tsp` file to control SDK generation. This term does not refer to modifications of the service API TypeSpec files.
```
+1

SDK Customization are customization that couldn't be expressed in client.tsp, and have been done in the target language. Those are always located in the language repo, along the package. They may require pre-generation of some constructs through client.tsp.

samvaity replied:
@chrisradek this spec addresses a gap: it focuses on automating spec-level customizations (renames, client splits, etc.), whereas the existing TspClientUpdateTool deals with code-level customizations. To achieve a seamless inner-loop, these must integrate into one unified workflow where the CLI handles both updating the client.tsp and updating any hand-written SDK code in tandem. 

Example: If a .NET analyzer error says “Class X should end with ‘Client’”, the tool should prefer to add a decorator in client.tsp. After updating the spec and regenerating, but there in an error in the handwritten code (not solvable by spec changes), where the current code patch tool (TspClientUpdateTool) would handle it.

lmazuel commented on tools/azsdk-cli/docs/specs/3-customizing-client-tsp.spec.md:
```diff
@@ -0,0 +1,93 @@
+
+TypeSpec specs are written to describe service APIs. However, our team supports generating SDKs in no less than 5 languages, each differing in their idioms. The `client.tsp` file acts as a TypeSpec entry point for SDK emitters, applying language-specific customizations to the service spec that are then used by each respective SDK emitter.
+
+Service teams cannot be expected to understand the nuances of each SDK language and how to best customize the generated code for each. Applying changes also requires understanding how to use [Azure.ClientGenerator.Core](https://azure.github.io/typespec-azure/docs/libraries/typespec-client-generator-core/reference/).
```
Not sure it's the main argument for client.tsp
- First, customization can be cross-language, so you do it once and for all. That applies for client structuration, some naming, etc.
- Then, you can put language specific as needed

IOW, it's incorrect to say that client.tsp is only for language specific changes. It's for SDK specific changes, and some instructions can apply to several of them. We do support for instance modifiers like `!python` to say "applies to everything but Python", because cross-language modification is important.

This means that when using AI to write client.tsp, we may have an interesting challenge at some point if each AI wants to write in the language they own, and nothing else, like ending up with:

```
@clientName(Foo, "bar", "python")
@clientName(Foo, "bar", "js")
@clientName(Foo, "bar", "csharp")
@clientName(Foo, "bar", "java")
```

which ideally we avoid.

The opposite problem is true, if you are working on Java, the AI needs to understand that this line applies to Java right now:
```
@clientName(Foo, "bar", "!python")
```

Commit:
Add code-customizations definition, and add manually triggered processes as a way to get feedback for changes

samvaity commented on tools/azsdk-cli/docs/specs/3-customizing-client-tsp.spec.md:
```diff
@@ -0,0 +1,93 @@
+**Go**:
+The Go SDK has a [breaking changes](https://github.com/Azure/azure-sdk-for-go/blob/instruction/.github/prompts/go-sdk-breaking-changes-review.prompts.md) prompt that analyzes the generated SDK's changelog for specific patterns and maps those to `client.tsp` customizations where possible.
+_We don't currently have a clear step in our process for identifying breaking changes, but the flow for Go is: Emit SDK -> Analyze changelog & apply fixes -> Repeat._
+
```
The current spec focuses on client.tsp changes, but as a tooling there’s an implicit coordination needed between the Azure REST API spec repository (where client.tsp lives) and the language SDK repos (where custom code lives). 
I suggest we have a single integrated tool, effectively touching both: updating the spec file and the custom code. Ideally, the tool could output a pair of PRs – one to the azure-rest-api-specs with the TypeSpec changes, and one to the azure-sdk-for- with the updated code – or if using a monorepo approach, a single PR if all changes are in one place.

Currently, without the integrated tool concept the **order of operations** becomes crucial and the user will have to perform sequence execution of commands to get the desired code updates.

samvaity reviewed (commented):
The spec introduces updating the TypeSpec itself. Looking at it from customizations pov, there’s isn't a direct functional overlap (spec vs code), but there are potential redundancies and conflicts that we would need to manage:
1. Dual Fixing of Issues: If an API review requests “Rename class FooClient to FooServiceClient in all languages,” spec automation (e.g., @clientName in Spec #12577) should handle this before code generation. Otherwise, code patches may duplicate the effort or cause build failures. Coordinating spec fixes before code generation prevents redundant work and keeps design changes at the source.

2. Redundant Documentation or Guidelines: Both spec and code fixes use language guidelines (e.g., .NET class naming). If spec automation applies conventions first, code patchers won’t need to rename classes, streamlining the process and letting code updates focus on areas spec can't cover, like test code.

3. Workflow Confusion: Separate spec and code fixes can confuse developers about which tool to run and in what order, leading to unnecessary patches or incomplete updates. Merging workflows removes this ambiguity.

Commit:
incorporate feedback on providing client.tsp expertise and the impact

Commit:
update gaps and define key areas design proposal is targetting

Commit:
expand story around code customizations

Commit:
fix grammatical error

m-redding commented on tools/azsdk-cli/docs/specs/3-customizing-client-tsp.spec.md:
```diff
@@ -0,0 +1,116 @@
+- **AI agents are not experts on client customizations.** They make mistakes when applying customizations to `client.tsp`, even with access to the existing documentation.
+- **AI agents don't know when to suggest client customizations.** There is no mechanism in azsdk-cli today to steer AI agents to focus on applying `client.tsp` customizations.
+- **Nothing shared across languages.** Each language team has their own process for identifying and applying `client.tsp` customizations.
+- **Difficult to apply changes from outside the inner loop.** We lack a process for updating `client.tsp` from outside the azure-rest-api-specs repo.
```
+1 to this, it's been awkward in .NET trying to make adjustments without a local copy of the specs repo.

ronniegeraghty replied:
It's been my assumption that most of the local work requires both the Specs repo and 1-All the SDK Language repos to be cloned locally. 

@praveenkuttappan do we have tools that are currently creating PRs for repos? Or is that done by out tool invoking a pipeline that has the PR creation logic?

m-redding reviewed (approved):
Adding more context for .NET but agree with the explanation of the problem/gaps/state/etc

Commit:
update spec with details for gaps 1 and 3

Commit:
Merge remote-tracking branch 'upstream/main' into spec-client-tsp-draft

samvaity committed:
refresh customization spec for unified approach

samvaity committed:
update for customizationrequest param

chrisradek committed:
Merge pull request #5 from chrisradek/samvaity-enhancements

Refresh customization spec for unified approach

Commit:
Update workflow for CustomCodeUpdateTool

Commit:
minor formatting

Commit:
Merge remote-tracking branch 'upstream/main' into spec-client-tsp-draft

Commit:
update spec title

Commit:
less specific to client.tsp

swathipil commented on tools/azsdk-cli/docs/specs/3-customizing-client-tsp.spec.md:
```diff
@@ -0,0 +1,510 @@
+_We don't currently have a clear step in our process for identifying breaking changes, but the flow for Go is: Emit SDK -> Analyze changelog & apply fixes -> Repeat._
+
+**Python**:
+The Python SDK has [breaking changes](https://github.com/Azure/azure-sdk-for-python/tree/main/scripts/breaking_changes_checker) checks that may be resolvable via `client.tsp` customizations, but there is no automated process for identifying or applying these client.tsp customizations today.
```
We have a GH issue today to incorporate the breaking changes tool directly into the Python APIView parser, where breaking public API changes would show up in the diagnostics banner. I believe Travis mentioned that they plan on having APIView Copilot add comments to help resolve diagnostics (such as linting errors), so this could be part of the existing flow from reviews automatically.

Commit:
add specific scenarios for client.tsp customization that should work

samvaity committed:
Revise customization tool documentation and examples

Updated documentation to reflect changes in the customization tool and its usage scenarios. Added new sections for limitations, pipeline considerations, and examples for user prompts.

Commit:
update workflow diagram, add context tracking and classifier sections to proposal

samvaity committed:
Updated definitions and descriptions for TypeSpec and Code Customizations

samvaity committed:
Revise customization workflow documentation

Updated clarity on the activation criteria for Phase B and the summary of changes returned by the tool.

samvaity committed:
update for conservative approach points add

samvaity committed:
Simplify workflow diagram in documentation

samvaity committed:
Refine rationale for Phase B automation

samvaity committed:
Revise two-phase customization workflow

Updated documentation to clarify the customization workflow, challenges, and proposed solutions for TypeSpec and SDK code customizations. Enhanced sections on AI agent challenges, entry points, and the two-phase customization workflow.

samvaity committed:
Revise customization workflow documentation for narrow scope Phase B

lmazuel commented on tools/azsdk-cli/docs/specs/3-customizing-client-tsp.spec.md:
```diff
@@ -0,0 +1,581 @@
+
+   - **Scope** (<20 lines, <5 files, deterministic only):
+     - ✅ **In**: Remove duplicates, update references, add imports, rename keywords, update type annotations
+     - ❌ **Out**: Convenience methods, architecture changes, visibility (use `@access`), error handling, complex logic
```
Curious how this list was built?

samvaity commented on tools/azsdk-cli/docs/specs/3-customizing-client-tsp.spec.md:
```diff
@@ -0,0 +1,581 @@
+  "tool": "azsdk_package_customize_code",
+  "arguments": {
+    "packagePath": "/path/to/sdk",
+    "customizationRequest": "Build failed with: error CS0246: The type or namespace name 'FooModel' could not be found"
```
The package path includes the language indicator, which is how we currently extract language information for our existing tools in MCP.

samvaity commented on tools/azsdk-cli/docs/specs/3-customizing-client-tsp.spec.md:
```diff
@@ -0,0 +1,581 @@
+
+| Phase                 | Action                                                                                                                        | Result                                                              |
+| --------------------- | ----------------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------- |
+| **Phase A: TypeSpec** | Analyze feedback requirements<br/>Apply `@clientName` with proper scoping for .NET<br/>Regenerate .NET SDK<br/>Validate build | SDK regenerates successfully<br/>Build passes<br/>No Phase B needed |
```
That's an interesting question. I believe this would be outside the scope of the current specification. Builds can fail if they reference an old model name and need updates, especially if the samples were manually written instead of generated. Generated samples should be covered, but in general, excluding samples from the build might require manual updates or using `azsdk_package_samples_generate` to regenerate them. So we can assume, sample updates are not part of the customization workflow.  I will add it to the scope.

samvaity committed:
Handle failure and add sample update note

samvaity committed:
Revise Scenario 7 for feature request and customization guidance

samvaity committed:
Handle failure, add sample update note, scenario 7 for manual guidance

chrisradek commented on tools/azsdk-cli/docs/specs/3-customizing-client-tsp.spec.md:
```diff
@@ -0,0 +1,621 @@
+## Success Criteria
+
+- [ ] **Automated Detection**: Correctly identify build failure types across all supported languages
+- [ ] **Phase A Success (80% Coverage)**: Successfully apply [TypeSpec Customizations](#typespec-customizations) for 80%+ of feature requests and specification issues
```
With telemetry we can track if client.tsp changes were made, and if the tool succeeds due to SDK regeneration/rebuild passing. We haven't discussed tracking actual diffs yet - that could be interesting to let us see if _we_ think the changes look good overtime too.

samvaity commented on tools/azsdk-cli/docs/specs/3-customizing-client-tsp.spec.md:
```diff
@@ -0,0 +1,621 @@
+
+1. **Phase B – [Code Customizations](#code-customizations):**
+
+   - **Activation**: Phase A build fails AND customization files exist (Java: `/customization/` or `*Customization.java`, Python: `*_patch.py`, .NET: partial classes)
```
For .NET, it would be  any partial classes exist outside of Generated folder.
